pkgs/development/libraries/flutter-engine/package.nix

```diff
@@ -0,0 +1,182 @@
+    installPhase = ''
+      mkdir -p $out
+      cp -r -P --no-preserve=mode,ownership $buildroot $out/src
+      cp -r -P --no-preserve=mode,ownership $src $out/src/flutter
+
+      ${concatStringsSep "\n" (attrValues (mapAttrs (name: src: ''
+        if [[ -d ${src} ]]; then
+          mkdir -p $out/${name}
+          cp -r -P --no-preserve=mode,ownership ${src}/* $out/${name}
+          find ${src} -type f -name '.*' | xargs -I {} cp --no-preserve=mode,ownership {} $out/${name}
+          ${optionalString (builtins.hasAttr "rev" src) ''
+            mkdir -p $out/${name}/.git/logs
+            echo "${src.rev}" >$out/${name}/.git/logs/HEAD
+          ''}
+        else
+          cp -P --no-preserve=mode,ownership ${src} $out/${name}
+        fi
+      '') flutter-deps))}
+
+      touch $out/src/third_party/dart/build/config/gclient_args.gni
+      cp $gitrev $out/src/flutter/build/git_revision.py
+
+      for bin in $binaryFixes; do
+        chmod 0755 $out/$bin
+        patchelf --set-interpreter ${stdenv.cc.libc}/lib/${interpreter} $out/$bin
+      done
+
+      for sh in $shellFixes; do
+        chmod 0755 $out/$sh
+        patchShebangs $out/$sh
+      done
+
+      chmod +x $out/src/build/linux/sysroot_ld_path.sh
+
+      sed -i '1 s|^.*$|#!${gnumake}/bin/make -f|' $out/src/third_party/harfbuzz/src/update-unicode-tables.make
+
+      mkdir -p $out/src/build/linux/debian_sid_${toolchainArch}-sysroot
+      tar -xf $toolchain -C $out/src/build/linux/debian_sid_${toolchainArch}-sysroot
+
+      cd $out
+      mkdir -p $out/src/third_party/dart/.dart_tool
+      cp -r --no-preserve=ownership,mode $dartPackageConfig $out/src/third_party/dart/.dart_tool/package_config.json
+
+      python3 $out/src/third_party/dart/tools/generate_package_config.py
+      python3 $out/src/third_party/dart/tools/generate_sdk_version_file.py
+      python3 $out/src/tools/remove_stale_pyc_files.py src/tools
+      python3 $out/src/flutter/tools/pub_get_offline.py
+    '';
+  };
+in stdenvNoCC.mkDerivation rec {
+  pname = "flutter-engine-${runtimeMode}";
+  version = builtins.substring 0 7 args.version;
+  inherit runtimeMode src;
+
+  nativeBuildInputs = [ git python3 pkg-config ninja stdenv.cc.libc ];
+
+  ## FIXME: create a sysroot toolchain for Nix and use it instead
+  ## TODO: generate for other platforms based on "targetPlatform"
+  configurePhase = ''
+    ./src/flutter/tools/gn \
+      --runtime-mode $runtimeMode \
+      --no-build-glfw-shell \
+      --prebuilt-dart-sdk \
+      --embedder-for-target \
+      --no-goma \
+      --out-dir $out/lib/flutter \
+      --target-dir $runtimeMode
+
+    mv $out/lib/flutter/out/$runtimeMode $out/lib/flutter/$runtimeMode
+    rmdir $out/lib/flutter/out
+  '';
+
+  ## TODO: add more builds and possibly better cross-compiling
+  buildPhase = ''
+    cd $out/lib/flutter/$runtimeMode
+
+    echo "Building flatc"
+    ninja flatc -j$NIX_BUILD_CORES
```

Comment:
with

```nix
{
    binaryFixes = [
      "src/third_party/dart/tools/sdks/dart-sdk/bin/dart"
      "src/flutter/third_party/gn/gn"
      "src/buildtools/linux-x64/clang/bin/clang"
      "src/buildtools/linux-x64/clang/bin/clang++"
    ];
```

ninja fails with

```
Building flatc
[88/88] LINK ./flatc
FAILED: flatc exe.unstripped/flatc 
../../../../../../../build/flutter-engine-src-857bd6b74c5eb56151bfafe91e7fa6a82b6fee25/src/buildtools/linux-x64/clang/bin/clang++ -Wl,--fatal-warnings -m64 -fPIC -Wl,-z,noexecstack -Wl,-z,now -Wl,-z,relro -Wl,-z,defs -pthread --sysroot=/build/flutter-engine-src-857bd6b74c5eb56151bfafe91e7fa6a82b6fee25/src/build/linux/debian_sid_amd64-sysroot -L/build/flutter-engine-src-857bd6b74c5eb56151bfafe91e7fa6a82b6fee25/src/build/linux/debian_sid_amd64-sysroot/usr/local/lib/x86_64-linux-gnu\ -Wl,-rpath-link=/build/flutter-engine-src-857bd6b74c5eb56151bfafe91e7fa6a82b6fee25/src/build/linux/debian_sid_amd64-sysroot/usr/local/lib/x86_64-linux-gnu\ -L/build/flutter-engine-src-857bd6b74c5eb56151bfafe91e7fa6a82b6fee25/src/build/linux/debian_sid_amd64-sysroot/lib/x86_64-linux-gnu\ -Wl,-rpath-link=/build/flutter-engine-src-857bd6b74c5eb56151bfafe91e7fa6a82b6fee25/src/build/linux/debian_sid_amd64-sysroot/lib/x86_64-linux-gnu\ -L/build/flutter-engine-src-857bd6b74c5eb56151bfafe91e7fa6a82b6fee25/src/build/linux/debian_sid_amd64-sysroot/usr/lib/x86_64-linux-gnu\ -Wl,-rpath-link=/build/flutter-engine-src-857bd6b74c5eb56151bfafe91e7fa6a82b6fee25/src/build/linux/debian_sid_amd64-sysroot/usr/lib/x86_64-linux-gnu -Wl,-O2 -Wl,--gc-sections -Wl,--as-needed -Wl,-rpath=\$ORIGIN/ -Wl,-rpath-link= -Wl,--disable-new-dtags  -o ./exe.unstripped/flatc -Wl,--build-id=sha1 -Wl,--start-group @./flatc.rsp  -Wl,--end-group  -ldl  && ../../../../../../../build/flutter-engine-src-857bd6b74c5eb56151bfafe91e7fa6a82b6fee25/src/buildtools/linux-x64/clang/bin/llvm-objcopy --strip-all ./exe.unstripped/flatc ./flatc
clang-15: error: unable to execute command: Executable "ld" doesn't exist!
```

binaries in flutter's clang:

```
ls $out/src/buildtools/linux-x64/clang/bin

analyze-build             ld64.lld              llvm-objdump
clang                     lld                   llvm-otool
clang++                   lld-link              llvm-pdbutil
clang-15                  llvm-ar               llvm-profdata
clang-apply-replacements  llvm-bolt             llvm-ranlib
clang-cl                  llvm-cov              llvm-rc
clang-cpp                 llvm-cxxfilt          llvm-readelf
clang-doc                 llvm-debuginfod-find  llvm-readobj
clang-format              llvm-dlltool          llvm-size
clang-include-fixer       llvm-dwarfdump        llvm-strip
clang-refactor            llvm-dwp              llvm-symbolizer
clang-scan-deps           llvm-gsymutil         llvm-undname
clang-tidy                llvm-ifs              llvm-xray
clangd                    llvm-lib              run-clang-tidy
dsymutil                  llvm-libtool-darwin   sancov
find-all-symbols          llvm-lipo             scan-build-py
git-clang-format          llvm-mt               wasm-ld
intercept-build           llvm-nm
ld.lld                    llvm-objcopy
```

also:
`ninja` should be `TERM=dumb ninja`

Reply:
I've resolved this locally.